### PR TITLE
Define product tests suites as code

### PR DESCRIPTION
**Pull request by wendigo**
Since we are defining test environments as code I think that we should do the same with product test suite definitions.

This PR changes suite definitions from bash scripts to Java code, tying environment implementation with test group definitions, exclusions and inclusions, and other configuration properties.

I've added `suite run` and `suite list` commands to the `presto-product-tests-launcher/bin/run-launcher`.

Usage:

```
presto-product-tests-launcher/bin/run-launcher suite run --suite suite-2 --suite-config config-cdh5

2020-08-20T16:55:55.981+0200	INFO	main	io.prestosql.tests.product.launcher.cli.SuiteRun	Starting suite 'suite-2' with config 'ConfigCdh5{hadoopBaseImage=Optional[prestodev/cdh5.15-hive], dockerImagesVersion=Optional[30], excludedGroups=[skip_on_cdh, iceberg], excludedTests=[], temptoEnvironmentConfigFile=Optional[/dev/null]}' and test runs:
2020-08-20T16:55:55.982+0200	INFO	main	io.prestosql.tests.product.launcher.cli.SuiteRun	 * environment 'singlenode': groups: [hdfs_no_impersonation, hive_compression], excluded groups: [], tests: [], excluded tests: []
2020-08-20T16:55:55.982+0200	INFO	main	io.prestosql.tests.product.launcher.cli.SuiteRun	 * environment 'singlenode-kerberos-hdfs-no-impersonation': groups: [storage_formats, hdfs_no_impersonation], excluded groups: [], tests: [], excluded tests: []
2020-08-20T16:55:55.982+0200	INFO	main	io.prestosql.tests.product.launcher.cli.SuiteRun	 * environment 'singlenode-hdfs-impersonation': groups: [storage_formats, cli, hdfs_impersonation], excluded groups: [], tests: [], excluded tests: []
2020-08-20T16:55:55.982+0200	INFO	main	io.prestosql.tests.product.launcher.cli.SuiteRun	 * environment 'singlenode-kerberos-hdfs-impersonation': groups: [storage_formats, cli, hdfs_impersonation, authorization, hive_file_header], excluded groups: [], tests: [], excluded tests: []
2020-08-20T16:55:55.983+0200	INFO	main	io.prestosql.tests.product.launcher.cli.SuiteRun	 * environment 'singlenode': groups: [hive_with_external_writes], excluded groups: [], tests: [], excluded tests: []
2020-08-20T16:55:55.983+0200	INFO	main	io.prestosql.tests.product.launcher.cli.SuiteRun	Running test run SuiteTestRun{environment=singlenode, groups=[hdfs_no_impersonation, hive_compression], excludedGroups=[], tests=[], excludedTests=[]} with config ConfigCdh5{hadoopBaseImage=Optional[prestodev/cdh5.15-hive], dockerImagesVersion=Optional[30], excludedGroups=[skip_on_cdh, iceberg], excludedTests=[], temptoEnvironmentConfigFile=Optional[/dev/null]}
2020-08-20T16:55:55.999+0200	INFO	main	io.prestosql.tests.product.launcher.cli.TestRun	Pruning old environment(s)
2020-08-20T16:55:55.999+0200	INFO	main	io.prestosql.tests.product.launcher.env.Environments	Shutting down previous containers
```

```
presto-product-tests-launcher/bin/run-launcher suite list
Available suites:
suite-1
suite-2
suite-3
suite-5
suite-6-non-generic
suite-7-non-generic
suite-8-non-generic
suite-tpcds

Available suite configuration profiles:
config-cdh5
config-empty
config-hdp3
```

TODO: Use test groups' constant names in a suite definition.

### CODE REVIEW

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/SuiteRun.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+            this.environmentFactory = requireNonNull(environmentFactory, "environmentFactory is null");
+        }
+
+        @Override
+        public void run()
+        {
+            Suite suite = suiteFactory.getSuite(suiteRunOptions.suite);
+            SuiteConfig suiteConfig = suiteFactory.getSuiteConfig(suiteRunOptions.config);
+
+            log.info("Starting suite '%s' with config '%s' and jobs: ", suite.getSuiteName(), suiteConfig);
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/SuiteRun.java:114]: use job or run consistently

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/SuiteRun.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        }
+
+        public void tryRunSuiteJob(SuiteTestRun suiteTestRun, SuiteConfig suiteConfig)
+        {
+            new TestRun.Execution(environmentFactory, pathResolver, createEnvironmentOptions(environmentOptions, suiteConfig), createTestRunOptions(suiteTestRun, suiteConfig)).run();
+        }
+
+        private EnvironmentOptions createEnvironmentOptions(EnvironmentOptions environmentOptions, SuiteConfig suiteConfig)
+        {
+            EnvironmentOptions copy = new EnvironmentOptions();
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/SuiteRun.java:159]: Create `EnvironmentOptions.copyOf` method and use here

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/SuiteConfig.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        return Optional.of(DEFAULT_HADOOP_BASE_IMAGE);
+    }
+
+    public List<String> getExcludedGroups()
+    {
+        return ImmutableList.of();
+    }
+
+    /**
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/SuiteConfig.java:41]: drop the comments like this one eventually.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/SuiteModule.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import static io.prestosql.tests.product.launcher.suite.Suites.nameForSuiteClass;
+import static java.util.Objects.requireNonNull;
+
+public final class SuiteModule
+        implements Module
+{
+    public static final String BASE_SUITES_PACKAGE = "io.prestosql.tests.product.launcher.suite";
+    private final Module additionalSuites;
+
+    public SuiteModule(Module additionalSuites)
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/SuiteModule.java:31]: Why do we need `additionalSuites`.

Can't we just do 
```
MapBinder<String, Suite> suites = newMapBinder(binder, String.class, Suite.class);
suites.addBinding(...)
```
in some other module if we need to extend the list?

**Review by losipiuk — COMMENTED** [review]
Nice :)

LGTM - minor comments.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/SuiteConfig.java (hunk 3 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/SuiteConfig.java:41]: I've kept it here for review to be simpler. I can drop them in a separate commit.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/SuiteRun.java (hunk 1 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/SuiteRun.java:114]: Good catch.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/SuiteConfig.java (hunk 3 above)
Comment (losipiuk) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/SuiteConfig.java:41]: Do you expect someone to actually read through all the definitions ? ;)

**Comment by findepi** [comment]
When a suite fails on CI, I need to understand which test run was the failing one (this isn't trivial today, but doable).
Then I run it locally (test run, not the whole suite), or modify parameters to run selected tests only.

How will this be facilitated now?

Same applies when developing a new test in case when it's not possible to run it from IDE. (cc @hashhar for his recent experience with this)

**Comment by wendigo** [comment]
> When a suite fails on CI, I need to understand which test run was the failing one (this isn't trivial today, but doable).

Yes, you will get exact information about failed test runs (actually you'll get a nice summary at the end of the suite). 
Each test is marked in the logs as it was when invoked from bash script. 

> Then I run it locally (test run, not the whole suite), or modify parameters to run selected tests only.
> 
> How will this be facilitated now?

You can still run the environment with tempo args as before. 

> 
> Same applies when developing a new test in case when it's not possible to run it from IDE. (cc @hashhar for his recent experience with this)

**Comment by findepi** [comment]
> > How will this be facilitated now?
> 
> You can still run the environment with tempo args as before.
> 

Sure. However, today i _know_ how to do it (copy & paste from suite script into my terminal).
How will I know that after the change?

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/SuiteTestRun.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+public class SuiteTestRun
+{
+    private static final String SKIP_TEST_ARGUMENT = "DISTRO_SKIP_TEST";
+    private static final String SKIP_GROUP_ARGUMENT = "DISTRO_SKIP_GROUP";
+
+    private static final String TEMPTO_TEST_ARG = "-t";
+    private static final String TEMPTO_GROUP_ARG = "-g";
+    private static final String TEMPTO_EXCLUDE_GROUP_ARG = "-x";
+    private static final String TEMPTO_EXCLUDE_TEST_ARG = "-t";
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/SuiteTestRun.java:37]: both test and exclude?

Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/SuiteTestRun.java]: woah, good catch 👍

**Comment by wendigo** [comment]
Passes CI so ready for final review @losipiuk @findepi

**Comment by wendigo** [comment]
@findepi I've added extra `suite describe` command:

```
Suite 'suite-3' with configuration 'config-hdp3' consists of following test runs: 

presto-product-tests-launcher/bin/run-launcher test run --hadoop-base-image prestodev/hdp3.1-hive \
--image-version 31 \
--hadoop-image-version 31 \
--tempto-environment-config-file /docker/presto-product-tests/conf/tempto/tempto-configuration-for-hive3.yaml \
--server-package presto-server/target/presto-server-341-SNAPSHOT.tar.gz \
--bind \
--test-jar presto-product-tests/target/presto-product-tests-341-SNAPSHOT-executable.jar \
--environment multinode-tls \
--reports-dir presto-product-tests/target/suite-3/config-hdp3/multinode-tls \
--startup-retries 5 \
-- -g smoke,cli,group-by,join,tls -x iceberg


presto-product-tests-launcher/bin/run-launcher test run --hadoop-base-image prestodev/hdp3.1-hive \
--image-version 31 \
--hadoop-image-version 31 \
--tempto-environment-config-file /docker/presto-product-tests/conf/tempto/tempto-configuration-for-hive3.yaml \
--server-package presto-server/target/presto-server-341-SNAPSHOT.tar.gz \
--bind \
--test-jar presto-product-tests/target/presto-product-tests-341-SNAPSHOT-executable.jar \
--environment multinode-tls-kerberos \
--reports-dir presto-product-tests/target/suite-3/config-hdp3/multinode-tls-kerberos \
--startup-retries 5 \
-- -g cli,group-by,join,tls -x iceberg


presto-product-tests-launcher/bin/run-launcher test run --hadoop-base-image prestodev/hdp3.1-hive \
--image-version 31 \
--hadoop-image-version 31 \
--tempto-environment-config-file /docker/presto-product-tests/conf/tempto/tempto-configuration-for-hive3.yaml \
--server-package presto-server/target/presto-server-341-SNAPSHOT.tar.gz \
--bind \
--test-jar presto-product-tests/target/presto-product-tests-341-SNAPSHOT-executable.jar \
--environment singlenode-kerberos-hdfs-impersonation-with-wire-encryption \
--reports-dir presto-product-tests/target/suite-3/config-hdp3/singlenode-kerberos-hdfs-impersonation-with-wire-encryption \
--startup-retries 5 \
-- -g storage_formats,cli,hdfs_impersonation,authorization -x iceberg


presto-product-tests-launcher/bin/run-launcher test run --hadoop-base-image prestodev/hdp3.1-hive \
--image-version 31 \
--hadoop-image-version 31 \
--tempto-environment-config-file /docker/presto-product-tests/conf/tempto/tempto-configuration-for-hive3.yaml \
--server-package presto-server/target/presto-server-341-SNAPSHOT.tar.gz \
--bind \
--test-jar presto-product-tests/target/presto-product-tests-341-SNAPSHOT-executable.jar \
--environment singlenode-kerberos-hdfs-impersonation-with-data-protection \
--reports-dir presto-product-tests/target/suite-3/config-hdp3/singlenode-kerberos-hdfs-impersonation-with-data-protection \
--startup-retries 5 \
-- -x iceberg -t TestHiveStorageFormats.testOrcTableCreatedInPresto,TestHiveCreateTable.testCreateTable
```

**Comment by wendigo** [comment]
CI failed https://github.com/prestosql/presto/issues/4885

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/SuiteTestRun.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+    private final Class<? extends EnvironmentProvider> environment;
+    private final List<String> groups;
+    private final List<String> excludedGroups;
+    private final List<String> excludedTests;
+    private final List<String> tests;
```
[the remark this answers is not in the document]
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/suite/SuiteTestRun.java]: and `suite describe` to list all test runs with corresponding configuration

**Comment by hashhar** [comment]
That `describe` command solves the concern Piotr raised (and I agree) in https://github.com/prestosql/presto/pull/4895#issuecomment-678122647 I think.

One change I can think of is to remove the test run command printed when a test starts from https://github.com/prestosql/presto/pull/4895#issuecomment-678363287 and instead print it only in case of a failed test run (if possible). Today it's difficult to identify which PT failed in a CI run because there's no easy term to grep for in the logs (or maybe there is but I don't know it). For the successful runs we already have the new `describe` command.

Also, how are we validating nothing got mistranslated from the shell script PTs to Java PTs? Are there any PT reports we can use to compare between both?

**Comment by wendigo** [comment]
Actually @hashhar, after suite is completed, there is a summary of test runs executed with the status attached:

```
Test runs summary (7 passed, 0 failed): 
2020-08-25T08:36:34.3313593Z 2020-08-25T08:36:34.330Z	INFO	main	io.prestosql.tests.product.launcher.cli.SuiteRun	 * 'SuiteTestRun{environment=singlenode-ldap, groups=[ldap], excludedGroups=[], tests=[], excludedTests=[]}' PASSED in 3.41m
2020-08-25T08:36:34.3314457Z 2020-08-25T08:36:34.330Z	INFO	main	io.prestosql.tests.product.launcher.cli.SuiteRun	 * 'SuiteTestRun{environment=singlenode-ldap-insecure, groups=[ldap], excludedGroups=[], tests=[], excludedTests=[]}' PASSED in 2.20m
2020-08-25T08:36:34.3315008Z 2020-08-25T08:36:34.330Z	INFO	main	io.prestosql.tests.product.launcher.cli.SuiteRun	 * 'SuiteTestRun{environment=singlenode-ldap-referrals, groups=[ldap], excludedGroups=[], tests=[], excludedTests=[]}' PASSED in 2.18m
2020-08-25T08:36:34.3315582Z 2020-08-25T08:36:34.330Z	INFO	main	io.prestosql.tests.product.launcher.cli.SuiteRun	 * 'SuiteTestRun{environment=singlenode-kerberos-kms-hdfs-no-impersonation, groups=[storage_formats], excludedGroups=[], tests=[], excludedTests=[]}' PASSED in 6.43m
2020-08-25T08:36:34.3316162Z 2020-08-25T08:36:34.330Z	INFO	main	io.prestosql.tests.product.launcher.cli.SuiteRun	 * 'SuiteTestRun{environment=singlenode-kerberos-kms-hdfs-impersonation, groups=[storage_formats], excludedGroups=[], tests=[], excludedTests=[]}' PASSED in 5.76m
2020-08-25T08:36:34.3316800Z 2020-08-25T08:36:34.330Z	INFO	main	io.prestosql.tests.product.launcher.cli.SuiteRun	 * 'SuiteTestRun{environment=singlenode-cassandra, groups=[cassandra], excludedGroups=[], tests=[], excludedTests=[]}' PASSED in 2.92m
2020-08-25T08:36:34.3317344Z 2020-08-25T08:36:34.330Z	INFO	main	io.prestosql.tests.product.launcher.cli.SuiteRun	 * 'SuiteTestRun{environment=singlenode-kafka, groups=[kafka], excludedGroups=[], tests=[], excludedTests=[]}' PASSED in 1.39m
```

**Comment by wendigo** [comment]
Other than that I've just checked if a number of test runs invocations are equal using sift (https://sift-tool.org/):

@master `sift 'run-launcher test run' -x sh | wc -l` 30
@this PR `sift 'testOnEnvironment(.*.class)' -x java | wc -l` 30

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentDefaults.java [hunk 7]
```diff
@@ -13,8 +13,6 @@
  */
 package io.prestosql.tests.product.launcher.env;
 
-import static java.lang.System.getenv;
-
 public class EnvironmentDefaults
 {
     public static final String DOCKER_IMAGES_VERSION = "31";
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentDefaults.java:18]: why do we need those getters after this change? Can't we just use constants?

**Review by losipiuk — COMMENTED** [review]
Mostly LGTM

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/SuiteDescribe.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+import java.io.FileOutputStream;
+import java.io.PrintStream;
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.Charset;
+
+import static io.prestosql.tests.product.launcher.cli.Commands.runCommand;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+@Command(name = "describe", description = "describe tests suite")
```
[the remark this answers is not in the document]
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/SuiteDescribe.java:44]: Sure.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/OptionsPrinter.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+import java.lang.reflect.Modifier;
+import java.util.List;
+
+import static java.lang.String.format;
+
+public final class OptionsPrinter
+{
+    private OptionsPrinter() {}
+
+    public static String toString(Object... objects)
```
[the remark this answers is not in the document]
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/OptionsPrinter.java:31]: Sure. I'll add unit for that.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/SuiteRun.java [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+            this.environmentOptions = requireNonNull(environmentOptions, "environmentOptions is null");
+            this.pathResolver = requireNonNull(pathResolver, "pathResolver is null");
+            this.suiteFactory = requireNonNull(suiteFactory, "suiteFactory is null");
+            this.environmentFactory = requireNonNull(environmentFactory, "environmentFactory is null");
+        }
+
+        @Override
+        public void run()
+        {
+            Suite suite = suiteFactory.getSuite(suiteRunOptions.suite);
```
[the remark this answers is not in the document]
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/SuiteRun.java]: Lack of logger's logger name and date - it makes copying content much easier (no need to remove prefix)

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/SuiteRun.java [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+            for (SuiteTestRun testRun : suiteTestRuns) {
+                results.add(executeSuiteTestRun(testRun, suiteConfig));
+            }
+
+            List<TestRunResult> testRunsResults = results.build();
+            boolean allSuccessful = testRunsResults.stream()
+                    .allMatch(TestRunResult::isSuccessful);
+
+            if (allSuccessful) {
+                log.info("Suite succeded in %s", nanosSince(suiteStartTime));
```
[the remark this answers is not in the document]
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/SuiteRun.java:156]: Actually it's airlift `Duration`'s `nanosSince` method which returns `Duration` which in turn has nice `toString()` implementation. So nanos are not displayed here 👍🏻

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentDefaults.java (hunk 7 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentDefaults.java:18]: Good point. Removing.

**Comment by wendigo** [comment]
@losipiuk addressed review 👍🏻

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentOptions.java [hunk 12]
```diff
@@ -15,25 +15,28 @@
...
+import io.prestosql.tests.product.launcher.PathResolver;
+import io.prestosql.tests.product.launcher.suite.SuiteConfig;
 
 import java.io.File;
 import java.util.Locale;
 
 import static com.google.common.base.MoreObjects.firstNonNull;
 import static java.util.Objects.requireNonNull;
 
-public final class EnvironmentOptions
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentOptions.java:25]: why so?

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/OptionsPrinter.java [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    {
+        List<String> arguments = stream(objects)
+                .map(OptionsPrinter::extractArguments)
+                .flatMap(Collection::stream)
+                .collect(toList());
+
+        return JOINER.join(arguments);
+    }
+
+    public static String format(Object object)
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/OptionsPrinter.java:46]: drop this one

**Review by losipiuk — APPROVED** [review]

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentOptions.java (hunk 12 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentOptions.java:25]: Sneaked pass the trenches. Sorry :)

**Comment by losipiuk** [comment]
please let me know after you rebase, and squash the commit which removes comments.

**Comment by wendigo** [comment]
@losipiuk done

**Comment by losipiuk** [comment]
https://github.com/prestosql/presto/pull/4891

**Comment by wendigo** [comment]
Woohoo :)

**Pull request merged**